WangscapeTest/TestCalculatorDither.cpp

```diff
@@ -0,0 +1,156 @@
+#include <gtest/gtest.h>
+#include <algorithm>
+#include <numeric>
+#include <tilegen/alpha/CalculatorDither.h>
+#include <boost/optional.hpp>
+
+class TestCalculatorDither : public ::testing::Test
+{
+public:
+    tilegen::alpha::CalculatorBase::Weights weight_examples;
+    std::vector<tilegen::alpha::CalculatorDither> cds;
+    tilegen::alpha::CalculatorDither& cd1;
+    tilegen::alpha::CalculatorDither& cd2;
+    tilegen::alpha::CalculatorDither& cd256;
+    std::mt19937 rng;
+    TestCalculatorDither() :
+        weight_examples{0., 1., 5678.8765, -1.},
+        cds(3),
+        cd1(cds[0]),
+        cd2(cds[1]),
+        cd256(cds[2])
+    {
+        std::random_device rd;
+        rng.seed(rd());
+        cd1.power = 1;
+        cd2.power = 2;
+        cd256.power = 256;
+    }
+    static ptrdiff_t winnerIndex(const tilegen::alpha::CalculatorBase::Alphas& alphas)
+    {
+        return std::distance(alphas.cbegin(), std::max_element(alphas.cbegin(), alphas.cend()));
+    }
+    static void validateResult(tilegen::alpha::CalculatorDither& cd,
+                               const tilegen::alpha::CalculatorBase::Weights& weights,
+                               boost::optional<std::pair<double, double>> expected_values = boost::none,
+                               boost::optional<size_t> expected_winners = 2)
+    {
+        std::set<ptrdiff_t> results;
+        for (int i = 0; i < 256; i++)
+        {
+            cd.updateAlphas(weights);
+            const auto& alphas = cd.getAlphas();
+            ptrdiff_t winner_index = winnerIndex(alphas);
+            results.insert(winner_index);
+            if (expected_values)
+            {
+                EXPECT_TRUE(weights[winner_index] == expected_values->first ||
+                            weights[winner_index] == expected_values->second);
+            }
+            for (size_t i = 0; i < alphas.size(); i++)
+            {
+                if (i == winner_index)
+                    EXPECT_EQ(255, alphas[i]);
+                else
+                    EXPECT_EQ(0, alphas[i]);
+            }
+        }
+        if (expected_winners)
+        {
+            if (cd.power < 10)
```

Comment:
This line separates `cd1` and `cd2` from `cd256`, which has very different behaviour. `10` could instead be anything from `2+epsilon` to `256`.